### PR TITLE
Add LinkTo component support to ANET entity links in editor

### LINKED ISSUE #1773: Epic 1773 - Improve rich text component
- [x] Currently, the rich text toolbar looks like a html editor. Users would be more comfortable with a toolbar that is similar to rich text editors. 
- [x] ~~Users have reported that spellchecking works in normal text fields but not in the rich text editor. Maybe it is a matter of simply turning `spellCheck` to `on` - worth investigating~~ https://github.com/facebook/draft-js/issues/640
- [ ] Users would like the ability to add references to other ANET objects in the body of the text (specifically other reports and people). We can experiment with adding a mention/link to anything that has a UUID in ANET (except for notes), possibly by using an advanced auto-suggest chooser. See #1811

### PR DESCRIPTION
**Pull request by jose-west**
Added the ability to find, parse and display links to ANET entities as LinkTo components in the text editor.

Fixes #1773 

### User changes
- Changes to the editor and text display fields in forms.

- [ ] anet.yml needs change
- [ ] db needs migration
- [ ] documentation has changed

### CODE REVIEW

File: client/src/components/LinkTo.js [hunk 1]
```diff
@@ -47,6 +47,8 @@ export default class LinkTo extends Component {
 
     target: PropTypes.string,
     whenUnspecified: PropTypes.string,
+    modelType: PropTypes.string,
+    model: PropTypes.object,
     style: PropTypes.object,
     ...modelPropTypes
```
Comment (gjvoosten) [inline client/src/components/LinkTo.js]: Why not use `model={…} modelType={…}` everywhere `LinkTo` is called instead of the (variable) prop named after the model? It would make the code cleaner (I never liked the fact that the props for `LinkTo` were not fixed/constant).

Comment (gjvoosten) [reply client/src/components/LinkTo.js]: I'll see if I can do a bulk-update myself.

File: client/src/utils_links.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  }
+`
+
+export const ENTITY_TYPES = {
+  REPORT: "report",
+  PERSON: "person",
+  ORGANIZATION: "organization",
+  POSITION: "position",
+  LOCATION: "anetLocation",
+  TASK: "task"
```
Comment (gjvoosten) [inline client/src/utils_links.js]: You should use the task label from the dictionary in the Settings.

File: client/src/components/editor/LinkSourceAnet.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+const widgetTypeMapping = {
+  [ENTITY_TYPES.REPORT]: widgetPropsReport,
+  [ENTITY_TYPES.PERSON]: widgetPropsPeople,
+  [ENTITY_TYPES.ORGANIZATION]: widgetPropsOrganization,
+  [ENTITY_TYPES.POSITION]: widgetPropsPosition,
+  [ENTITY_TYPES.LOCATION]: widgetPropsLocation,
+  [ENTITY_TYPES.TASK]: widgetPropsTask
+}
+
+class LinkSourceAnet extends Component {
```
Comment (gjvoosten) [inline client/src/components/editor/LinkSourceAnet.js]: This really ought to be a pure functional component.

File: client/src/models/Report.js [hunk 4]
```diff
@@ -288,6 +289,17 @@ export default class Report extends Model {
       )
   })
 
+  static autocompleteQuery =
+    "uuid, intent, engagementDate, duration, keyOutcomes, nextSteps, cancelledReason, atmosphere, author { uuid, name, rank, role }, tags { uuid, name, description }, updatedAt"
```
Comment (gjvoosten) [inline client/src/models/Report.js]: Do you really need _all_ these fields? Seeing as the template directly below only uses the intent and the author.

File: client/src/utils_links.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        return <LinkAnet url={domNode.attribs.href} />
+      }
+    }
+  })
+}
+
+export function getEntityByUuid(type, uuid) {
+  const entityQuery = parsedEntityLinkTypeQuery[type]
+  if (!entityQuery) {
+    console.log(`Unsupported entity type: ${type}`)
```
Comment (gjvoosten) [inline client/src/utils_links.js]: Don't log to console.

File: client/src/utils_links.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+  if (splittedUrl.length > 1) {
+    const typeRaw = splittedUrl[splittedUrl.length - 2]
+    const uuid = splittedUrl[splittedUrl.length - 1]
+    const type = parsedEntityLinkType[typeRaw]
+
+    if (type && new RegExp(UUID_REGEX).test(uuid)) {
+      return { type: type, uuid: uuid }
+    } else {
+      console.log(`Failed to parse entity type (${type}) or UUID (${uuid}).`)
```
Comment (gjvoosten) [inline client/src/utils_links.js]: Don't log to console.

File: client/src/components/LinkTo.js (hunk 1 above)
Comment (gjvoosten) [reply client/src/components/LinkTo.js]: Done.

File: client/src/components/RelatedObjectNotes.js [hunk 7]
```diff
@@ -438,13 +453,7 @@ class BaseRelatedObjectNotes extends Component {
...
                     )}
                   </div>
-                  <div
-                    style={{
-                      overflowWrap: "break-word",
-                      /* IE: */ wordWrap: "break-word"
-                    }}
-                    dangerouslySetInnerHTML={{ __html: noteText }}
-                  />
+                  {parseHtmlWithLinkTo(noteTextDiv)}
```
Comment (gjvoosten) [inline client/src/components/RelatedObjectNotes.js]: Can't this simply be:
```
                   <div
                    style={{
                      overflowWrap: "break-word",
                      /* IE: */ wordWrap: "break-word"
                    }}>
                     {parseHtmlWithLinkTo(noteText)}
                   </div>
```
so you don't need ReactDOMServer?

File: client/src/components/RichTextEditor.js [hunk 8]
```diff
@@ -100,6 +114,12 @@ const importerConfig = {
     }
 
     return null
+  },
+  htmlToStyle: (nodeName, node, currentStyle) => {
+    if (nodeName === "mark") {
```
Comment (gjvoosten) [inline client/src/components/RichTextEditor.js]: Shouldn't you also check for `"strike"` (counterpart of `styleToHTML`) or am I missing something?
(Assuming there were good reasons to add mark/strike.)

File: client/src/utils_links.js (hunk 2 above)
Comment (jose-west) [reply client/src/utils_links.js]: I think it would be a better alternative to use `getModelNameLinkTo` from the corresponding Model entity, as what I am trying to achieve is linking entity types to url links.

File: client/src/components/editor/LinkSourceAnet.js (hunk 3 above)
Comment (jose-west) [reply client/src/components/editor/LinkSourceAnet.js]: Indeed! I already changed the rest, but did not want to rework 'old work'. However, since I am busy I will also convert this one.

File: client/src/models/Report.js (hunk 4 above)
Comment (jose-west) [reply client/src/models/Report.js]: Good catch.

File: client/src/utils_links.js (hunk 6 above)
Comment (jose-west) [reply client/src/utils_links.js]: I just found that there is an `API.logonServer` call, so I will use that instead.

File: client/src/utils_links.js (hunk 2 above)
Comment (jose-west) [reply client/src/utils_links.js]: Apparently I was responding to an old commit. This comment was about labels, which should indeed come from the dictionary.

File: client/src/components/RichTextEditor.js (hunk 8 above)
Comment (jose-west) [reply client/src/components/RichTextEditor.js]: I was also surprised when I implemented it, but _mark_ is a case which is not supported by default by the editor (unlinke _strike_).

File: client/src/utils_links.js (hunk 2 above)
Comment (jose-west) [reply client/src/utils_links.js]: I actually ended up reusing the labels from SEARCH_OBJECT_LABELS (index.js) instead, as this dictionary was being use for 2 different purposes: labeling and object construction.

File: client/src/models/Report.js (hunk 4 above)
Comment (jose-west) [reply client/src/models/Report.js]: And `uuid` for row key in the table.

File: client/src/utils_links.js (hunk 6 above)
Comment (jose-west) [reply client/src/utils_links.js]: I ended up removing the logging and returning a null when no entiry could be found. This should be handled by the function rendering the link (in put a TODO)

File: client/src/components/RichTextEditor.js [hunk 9]
```diff
@@ -130,6 +150,16 @@ const exporterConfig = {
     }
 
     return originalText
+  },
+
+  styleToHTML: style => {
+    if (style === INLINE_STYLE.STRIKETHROUGH) {
+      return <strike />
+    }
```
Comment (gjvoosten) [inline client/src/components/RichTextEditor.js]: So by your comment https://github.com/NCI-Agency/anet/pull/2790/files#r375806248 , it this then still needed?

Comment (jose-west) [reply client/src/components/RichTextEditor.js]: Indeed, this check is obsolete and can be safely removed. It comes from a previous version of the editor which required dedicated HMTL style parsing.

Comment (jose-west) [reply client/src/components/RichTextEditor.js]: Just to be clear: `<mark>` is supported out of the box (both ways). `<strike>` needs parsing when converting the editor style to HMTL.

File: client/src/components/editor/Icon.js [hunk 10]
```diff
@@ -24,9 +24,9 @@ const Icon = ({ icon, title, className }) => {
 
   return (
     <svg
-      width="16"
-      height="16"
-      viewBox="0 0 1024 1024"
+      width={width || "16"}
+      height={height || "16"}
+      viewBox={viewBox || "0 0 1024 1024"}
```
Comment (gjvoosten) [inline client/src/components/editor/Icon.js]: Make these props required and set a default.

File: client/src/components/editor/LinkSource.js [hunk 11]
```diff
@@ -46,7 +46,7 @@ class LinkSource extends Component {
     if (url) {
       const selection = editorState.getSelection()
       const entityData = {
-        url: url.replace(/\s/g, "")
+        url: url
```
Comment (gjvoosten) [inline client/src/components/editor/LinkSource.js]: Why this change (or: why were spacced stripped before)?

File: client/src/components/editor/LinkSource.js [hunk 12]
```diff
@@ -55,7 +55,7 @@ class LinkSource extends Component {
         const contentState = editorState.getCurrentContent()
         const contentStateWithEntity = contentState.createEntity(
           entityType.type,
-          "MUTABLE",
+          "IMMUTABLE",
```
Comment (gjvoosten) [inline client/src/components/editor/LinkSource.js]: It makes sense to make ANET links immutable ( https://www.draftail.org/docs/content-storage#entities ), however, it would appear I am now no longer able to create abritrary (URL) links (except by pasting a URL copied from sonewhere else, which I then can no longer edit).

File: client/src/utils_links.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    const uuid = splittedUrl[splittedUrl.length - 1]
+    const type = parsedEntityLinkType[typeRaw]
+
+    if (type && new RegExp(UUID_REGEX).test(uuid)) {
+      return { type: type, uuid: uuid }
+    }
+  }
+
+  return null
+}
```
Comment (gjvoosten) [inline client/src/utils_links.js]: This function appears to rely heaviliy on the defined routing in `Routing.js`.

File: client/src/utils_links.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+  return API.query(entityQuery, {
+    uuid: uuid
+  }).then(data => {
+    const entity =
+      data.report ||
+      data.person ||
+      data.organization ||
+      data.position ||
+      data.location ||
+      data.task
```
Comment (gjvoosten) [inline client/src/utils_links.js]: This can be simplified by giving the requested data in the GQL definition an alias, e.g.:
```
const GQL_GET_REPORT = gql`
  query($uuid: String!) {
    entity: report(uuid: $uuid) {
      uuid
      intent
    }
  }
`
```
The you can just use `data.entity` if you use the same alias in GQL defs. I'll push an update for this.

File: client/src/components/editor/LinkSourceAnet.js [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        </ButtonToggleGroup>
+      </Modal.Body>
+
+      <Modal.Footer>
+        <AdvancedSingleSelect
+          ref={advancedSelectRef}
+          autofocus="true"
+          fieldName="entitySelect"
+          fieldLabel="Search in ANET:"
+          placeholder={"Find" + objectType.toLowerCase()}
```
Comment (gjvoosten) [inline client/src/components/editor/LinkSourceAnet.js]: Missing a space, should be `"Find " + …`.

File: client/src/components/editor/Icon.js (hunk 10 above)
Comment (jose-west) [reply client/src/components/editor/Icon.js]: I agree about setting a default instead of the current implementation, but why make them required?

File: client/src/components/editor/LinkSource.js (hunk 12 above)
Comment (jose-west) [reply client/src/components/editor/LinkSource.js]: In the previous implementation you were only allowed to edit links through a menu (when the link was clicked on in the editor). However, since that is gone I guess we should make them mutable, so we can edit them without copy/pasting.

File: client/src/utils_links.js (hunk 13 above)
Comment (jose-west) [reply client/src/utils_links.js]: Yes, indeed. Are you suggesting linking the entity address url links in `Routing.js` to the enitity names used in `parsedEntityLinkType`?

File: client/src/components/editor/Icon.js (hunk 10 above)
Comment (gjvoosten) [reply client/src/components/editor/Icon.js]: You're right, add a default (then they won't be required).

File: client/src/utils_links.js (hunk 13 above)
Comment (gjvoosten) [reply client/src/utils_links.js]: Not sure yet what the best solution is, but something that uses a single source of truth for both the routing paths and the URL splitting would be the ultimate goal.

File: client/src/components/editor/LinkSource.js (hunk 12 above)
Comment (gjvoosten) [reply client/src/components/editor/LinkSource.js]: On `candidate`, I can edit the text *and* the corresponding link independently.

Comment (gjvoosten) [reply client/src/components/editor/LinkSource.js]: For ANET links, it does not make sense to be able to edit the text, since it should be derived from the link (by whatever LinkTo comes up with). But for plain URL's, I'd like the old behaviour. That would mean deciding what the type of link is and do different things with it (for starters, one would be immutable, the other mutable). Would it be possible to add a new option/button to the pop-up modal dialog for abitrary URL's?

File: client/src/utils_links.js (hunk 13 above)
Comment (jose-west) [reply client/src/utils_links.js]: I think that a dictionary/object holding _page <--> link URL_ (in `Routing.js`) could serve as the single place of truth. The function above can refer to it for parsing (with an object mapping _entity <--> entity page_).

**Comment by gjvoosten** [comment]
@jose-west File `client/src/components/editor/LinkSource.js` is no longer used, remove it?

File: client/src/components/editor/LinkSource.js (hunk 12 above)
Comment (jose-west) [reply client/src/components/editor/LinkSource.js]: @VassilIordanov commented the following on this issue: 

> Jose, I do not think we should have external links (at least I do not want to encourage the users of doing that)

The editor does not support 2 types of custom links, so that is why the popover link editor is gone (but not removed from the code, see `LinkSource.js` and `Link.js`). I think we should remove link support entirely for non ANET entities, as it is confusing and undesired. In the current version (`candidate` branch) when you type out a link it does not parse it as such.

Comment (gjvoosten) [reply client/src/components/editor/LinkSource.js]: OK, so only ANET links then. About obsolete files, see https://github.com/NCI-Agency/anet/pull/2790#issuecomment-585617187 .

File: client/src/components/editor/LinkSource.js (hunk 11 above)
Comment (jose-west) [reply client/src/components/editor/LinkSource.js]: This comment applies to a file no longer in use.

File: client/src/components/editor/LinkSource.js (hunk 12 above)
Comment (jose-west) [reply client/src/components/editor/LinkSource.js]: I dropped all support for links not pointing to a ANET entity, but you can now paste a ANET entity link in the editor and it will be parsed to a correct `LintkTo` component on save.

File: client/src/components/editor/MediaBlock.js [hunk 16]
```diff
@@ -1,111 +0,0 @@
-// @flow
-import React, { Component } from "react"
-import PropTypes from "prop-types"
-import Tooltip from "components/Tooltip"
-import Portal from "components/Portal"
```
Comment (maradragan) [inline client/src/components/editor/MediaBlock.js]: As MediaBlock and TooltipEntity have been removed, I think components/Portal.js can also be removed, it is not being used somewhere else.

Comment (jose-west) [reply client/src/components/editor/MediaBlock.js]: Correct, and thanks to your remark I just noticed it was  not the only obsolete file.

File: client/src/models/Report.js [hunk 17]
```diff
@@ -288,6 +288,16 @@ export default class Report extends Model {
       )
   })
 
+  static autocompleteQuery = "uuid, intent, author { uuid, name, rank, role }"
+
+  static autocompleteTemplate(report) {
```
Comment (maradragan) [inline client/src/models/Report.js]: I think this is not being used anywhere.

File: client/src/components/editor/LinkAnetEntity.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+const LinkAnetEntity = ({ type, uuid, children }) => {
+  const [entity, setEntity] = useState()
+
+  useEffect(() => {
+    const response = getEntityByUuid(type, uuid)
+    if (response) {
+      response.then(data => setEntity(data))
+    }
+  }, [type, uuid])
```
Comment (maradragan) [inline client/src/components/editor/LinkAnetEntity.js]: You miss here the getEntityByUuid dependency.

Comment (maradragan) [reply client/src/components/editor/LinkAnetEntity.js]: Wait, I see lint-fix doesn't add it neither, I guess it's because it's just an import and not a prop or something else which changes.

File: client/src/components/editor/Icon.js [hunk 19]
```diff
@@ -24,13 +23,13 @@ const Icon = ({ icon, title, className }) => {
 
   return (
     <svg
-      width="16"
-      height="16"
-      viewBox="0 0 1024 1024"
+      width={width}
+      height={height}
+      viewBox={viewBox}
       className={`Draftail-Icon ${className || ""}`}
```
[the remark this answers is not in the document]
Comment (jose-west) [reply client/src/components/editor/Icon.js]: Good point.

File: client/src/models/Report.js (hunk 17 above)
Comment (jose-west) [reply client/src/models/Report.js]: It is used to retrieve the report information needed for the auto-complete when searching ANET entities.

File: client/src/components/editor/LinkAnetEntity.js (hunk 18 above)
Comment (jose-west) [reply client/src/components/editor/LinkAnetEntity.js]: I am afraid I am not following you here. Line 4 does import this dependency (`import { getEntityByUuid } from "utils_links"`), or am I missing something?

File: client/src/components/editor/LinkSourceAnet.js [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+
+    onComplete(nextState)
+  }
+
+  function changeObjectType(newObjectType) {
+    setObjectType(newObjectType)
+    setAdvancedSelectProps(widgetTypeMapping[newObjectType])
+
+    // Filter and type changed, need to update search results
+    advancedSelectRef.current.refreshSearch()
```
Comment (maradragan) [inline client/src/components/editor/LinkSourceAnet.js]: What's the point of doing this? When you change type the overlay gets closed, when you focus the input it will retrieve new results anyway. But maybe I'm missing something?

File: client/src/models/Report.js (hunk 17 above)
Comment (maradragan) [reply client/src/models/Report.js]: As far as I can see it it is no longer used. It is only defined in all models, but not being used.

Comment (jose-west) [reply client/src/models/Report.js]: Try to remove the line and serach for a report, you will see that it fails to fetch the data.

File: client/src/components/editor/LinkSourceAnet.js (hunk 20 above)
Comment (jose-west) [reply client/src/components/editor/LinkSourceAnet.js]: It used to be stuck after changing the type filter, so a refresh was needed, but I am not observing that behaviour anymore. Removed.

File: client/src/models/Report.js (hunk 17 above)
Comment (gjvoosten) [reply client/src/models/Report.js]: The `autocompleteTemplate` should no longer be used anywhere.

Comment (gjvoosten) [reply client/src/models/Report.js]: And `Report.autocompleteQuery` is also unused.

Comment (maradragan) [reply client/src/models/Report.js]: autocompleteTemplate is really not being used, I removed it and I can still search for reports. I do get in the console the error: "/files/mara/workspace-ncia-anet/anet/client/src/models/Report.js
  10:8  error  'React' is defined but never used  no-unused-vars". So that one needs to be removed too, maybe that's why it didn't work for you.

Comment (jose-west) [reply client/src/models/Report.js]: Indeed.

File: client/src/components/editor/LinkSourceAnet.js (hunk 20 above)
Comment (maradragan) [reply client/src/components/editor/LinkSourceAnet.js]: I see you didn't remove it, but I tested better and you are right.
Without refreshsearch it doesn't work properly, without the refreshSearch:
- I got to the Anet entity widget in the rich text editor
- I click on People, focus on the input field
- I select All advisors filter type, I see the advisors
- I than click on Organizations, focus on the input field
- I see just as many results as before, and all with empty names
- in the console I see the error that filterDefs is not defined.

So it's ok like this.
